synthtool/languages/node_mono_repo.py

```diff
@@ -0,0 +1,397 @@
+
+
+def install(hide_output=False):
+    """
+    Installs all dependencies for the current Node.js library.
+    """
+    logger.debug("Installing dependencies...")
+    shell.run(["npm", "install"], hide_output=hide_output)
+
+
+def fix(hide_output=False):
+    """
+    Fixes the formatting in the current Node.js library.
+    Before running fix script, run prelint to install extra dependencies
+    for samples, but do not fail if it does not succeed.
+    """
+    logger.debug("Running prelint...")
+    shell.run(["npm", "run", "prelint"], check=False, hide_output=hide_output)
+    logger.debug("Running fix...")
+    shell.run(["npm", "run", "fix"], hide_output=hide_output)
+
+
+def fix_hermetic(hide_output=False):
+    """
+    Fixes the formatting in the current Node.js library. It assumes that gts
+    is already installed in a well known location on disk:
+    """
+    logger.debug("Copy eslint config")
+    shell.run(
+        ["cp", "-r", f"{_TOOLS_DIRECTORY}/node_modules", "."],
+        check=True,
+        hide_output=hide_output,
+    )
+    logger.debug("Running fix...")
+    shell.run(
+        [f"{_TOOLS_DIRECTORY}/node_modules/.bin/gts", "fix"],
+        check=False,
+        hide_output=hide_output,
+    )
+
+
+def compile_protos(hide_output=False):
+    """
+    Compiles protos into .json, .js, and .d.ts files using
+    compileProtos script from google-gax.
+    """
+    logger.debug("Compiling protos...")
+    shell.run(["npx", "compileProtos", "src"], hide_output=hide_output)
+
+
+def compile_protos_hermetic(hide_output=False):
+    """
+    Compiles protos into .json, .js, and .d.ts files using
+    compileProtos script from google-gax.
+    """
+    logger.debug("Compiling protos...")
+    shell.run(
+        [f"{_TOOLS_DIRECTORY}/node_modules/.bin/compileProtos", "src"],
+        check=True,
+        hide_output=hide_output,
+    )
+
+
+def postprocess_gapic_library(hide_output=False):
+    logger.debug("Post-processing GAPIC library...")
+    install(hide_output=hide_output)
+    fix(hide_output=hide_output)
+    compile_protos(hide_output=hide_output)
+    logger.debug("Post-processing completed")
+
+
+def postprocess_gapic_library_hermetic(hide_output=False):
+    logger.debug("Post-processing GAPIC library...")
+    fix_hermetic(hide_output=hide_output)
+    compile_protos_hermetic(hide_output=hide_output)
+    logger.debug("Post-processing completed")
+
+
+default_staging_excludes = ["README.md", "package.json", "src/index.ts"]
+default_templates_excludes: List[str] = []
+
+
+def _noop(library: Path) -> None:
+    pass
+
+
+# 1. Copy all the staging files at the top level of our monorepo into our subfolder
+def walk_through_owlbot_dirs(dir: Path):
+    """
+    Walks through samples/generated to find all snippet metadata files, appends them to a list
+
+    Returns:
+    A list of all metadata files.
+    """
+    owlbot_dirs = []
+    for path_object in dir.glob("**/*"):
```

Comment:
I think there's a risk that this glob will get slow once we have dozens of repositories in the repo, as it's going to just create a list of all files in the entire mono-repo.

By convention, we're always going to have the `.OwlBot.yaml` in one location:

```
packages/
   a/.OwlBot.yaml
   b/.OwlBot.yaml
   b/.OwlBot.yaml
```

Given this, I would just use `os.listdir`, to list directories in `packages/`, and then check for a `.OwlBot.yaml` in each folder.